src/Sites/app/Runtime/TabCompleteDetection.tsx

```diff
@@ -96,15 +92,16 @@ export class TabCompleteDetection {
 				this.handleTabPress(ev, tabValues);
 			}
 		};
-		input.addEventListener('keydown', this.keyListener, {
-			capture: true
-		});
-
 		this.finalizeListener = () => this.resetCursor();
-		input.addEventListener('change', this.finalizeListener, {
-			capture: false
-		});
+		let chatRoomContent = document.querySelector(Twitch.Selectors.ChatRoomContent);
 
+		if(chatRoomContent) {
+			const observer = new MutationObserver(() => {
+				this.updateInput();
+			});
+			observer.observe(chatRoomContent, {subtree: false, childList: true});
```

Comment:
Mutation observers can be pretty expensive. I believe there should be a way to hook into Twitch's react to detect the timeout states

Reply:
Thanks, I'll investigate further.

Reply:
Never mind the previous comment. I found the ban status here `ChatComponent.props.currentUserBannedStatusData?.channel?.self?.banStatus?.expiresInMs` and added a check for update every 1s.

Reply:
This isn't really what I'm looking for either, unfortunately I am busy with the ongoing work on API V3 and the new website so i cannot guide you on this - but the ideal sitution would be to hook into the exact states that cause the timeout screen to appear/disappear.

Reply:
When you get timedout you get sent a gql request with the operation name `ChatRestrictions`

The only way I see doing this without checking `ChatComponent.props.currentUserBannedStatusData?.channel?.self?.banStatus?.expiresInMs` constantly is by [intercepting these request bodys](https://medium.com/@ddamico.125/intercepting-network-response-body-with-a-chrome-extension-b5b9f2ef9466) In this request `[0].data.channel.self.banStatus.isPermanent` will say whether or not the ban is permanent (if it's a timeout or ban). If this is you can check `ChatComponent.props.currentUserBannedStatusData?.channel?.self?.banStatus?.expiresInMs` since it for some reason, from what I can see doesn't include an expiry (weird).From there you can approach it however you'd like.

Snooping around twitch uses, to an extent, the way Cameron said except they don't check every second (obviously) and it's called from somewhere somehow idk, but I'm guessing it's something to do with the GQL request.

Reply:
Hey @AnatoleAM I think I've figured out what you're looking for. This should only trigger when the chat component is updated because of a change to the ban status prop (which removes/adds the chat input).

I understand this is probably a low priority issue. If there is something higher priority that I can work on, I am more than happy to help! (I'm fimiliar with front and back end code)

Reply:
That looks better, I'll have a look and test it when I can. Thank you for the fast iterations